Add better build instructions to the README.

felddy opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

Closes #50 

* Add instructions for cross-platform builds
* Clean up shields
* Clean up tables for each Docker configuration item
* Add section about supported tags

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

It was pointed out that the instructions for cross platform builds were: lacking.

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

Tested locally and in CI.

## 📷 Screenshots (if appropriate) ##

<!-- Remove this section and header if not needed -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

felddy committed:
Add better build instructions to the README.

* Add instructions for cross-platform builds
* Clean up shields
* Clean up tables for each Docker configuration item
* Add section about supported tags

jsf9k reviewed (approved):
Definitely helps.  Thank you.

dav3r reviewed (approved):
Lovely!  Nice improvements all around. 🥏

mcdonnnj commented on README.md:
```diff
@@ -1,8 +1,8 @@
 # skeleton-docker 💀🐳 #
 
-[![GitHub Build Status](https://github.com/cisagov/skeleton-docker/workflows/build/badge.svg)](https://github.com/cisagov/skeleton-docker/actions)
-[![Total alerts](https://img.shields.io/lgtm/alerts/g/cisagov/skeleton-docker.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/cisagov/skeleton-docker/alerts/)
-[![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/cisagov/skeleton-docker.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/cisagov/skeleton-docker/context:python)
```
If we're getting rid of LGTM badges, should we remove LGTM analysis? The tests written in Python are still being analyzed.

felddy replied:
You are correct: we should also do this in a PR to `skeleton-python-library`.

mcdonnnj commented on README.md:
```diff
@@ -1,8 +1,8 @@
-[![Total alerts](https://img.shields.io/lgtm/alerts/g/cisagov/skeleton-docker.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/cisagov/skeleton-docker/alerts/)
-[![Language grade: Python](https://img.shields.io/lgtm/grade/python/g/cisagov/skeleton-docker.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/cisagov/skeleton-docker/context:python)
+[![GitHub Build Status](https://github.com/cisagov/skeleton-docker/workflows/build/badge.svg)](https://github.com/cisagov/skeleton-docker/actions/workflows/build.yml)
+[![CodeQL](https://github.com/cisagov/skeleton-docker/workflows/CodeQL/badge.svg)](https://github.com/cisagov/skeleton-docker/actions/workflows/codeql-analysis.yml)
```
Just a thought so it's not reliant on a specific filename.

```suggestion
[![CodeQL](https://github.com/cisagov/skeleton-docker/workflows/CodeQL/badge.svg)](https://github.com/cisagov/skeleton-docker/actions?query=workflow%3ACodeQL)
```

felddy replied:
This ties the URL to the `name` field within a workflow.  I'm not sure that it is less fragile. 
It also prevents the nifty widget highlight from working in the Actions dashboard.
And it omits 6 workflow runs.  Which 6?!  I do not know:

# Filename
![Screen Shot 2021-04-08 at 14 41 11](https://user-images.githubusercontent.com/3229435/114080339-4a4da780-9879-11eb-9505-82f420c3a84d.png)

# Query name
![Screen Shot 2021-04-08 at 14 40 56](https://user-images.githubusercontent.com/3229435/114080349-4e79c500-9879-11eb-93af-438c45c6a34d.png)

mcdonnnj commented on README.md:
````diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+1. Then add the secret to your `docker-compose.yml` file:
+
+    ```yaml
+    version: "3.8"
````
Do either examples include any `3.8` features? We should generally use the lowest compose version we can for compatibility.

felddy replied:
How far back do you want to go?  
When do we move it forward?

https://docs.docker.com/compose/compose-file/compose-versioning/

mcdonnnj commented on README.md:
```diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+
+## Updating ##
+
+### Docker-compose ###
```
```suggestion
### Docker Compose ###
```

hillaryj replied:
To add context for anyone else who might search - Docker Compose is the name of the tool so this follows our style guide. ✅

mcdonnnj commented on README.md:
```diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+
+### Docker-compose ###
+
+1. Pull the new image from Docker hub:
```
In the Docker directions you specify stopping the container so should there be a `docker-compose down` in the compose section?

hillaryj replied:
Yes, I recommend an suggested first step of:


> 1. If your container is running, stop it before updating:
>
> ```console 
> docker-compose down
> ```

felddy replied:
docker-compose is a bit more magical.  It will determine which services need to be brought down and back up based on which images have been updated, and their dependents.  So a `pull` and `up` will "do the right thing™". 

https://docs.docker.com/compose/reference/pull/

https://docs.docker.com/compose/reference/up/

> If there are existing containers for a service, and the service’s configuration or image was changed after the container’s creation, `docker-compose up` picks up the changes by stopping and recreating the containers (preserving mounted volumes). To prevent Compose from picking up changes, use the `--no-recreate` flag.

hillaryj replied:
mmmmmagic 👍🏻  coolio

mcdonnnj commented on README.md:
````diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+    docker-compose pull
+    ```
+
+1. Recreate the running container:
````
In the Docker section you point toward the previous running instructions, should we do the same for compose?

hillaryj replied:
```suggestion
1. Recreate the running container by following the [previous instructions](#running-with-docker-compose):
```

mcdonnnj commented on README.md:
```diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+docker build \
+  --build-arg VERSION=0.0.1 \
+  --tag cisagov/example:0.0.1 \
+  https://github.com/cisagov/example.git#develop
```
👍

mcdonnnj reviewed (approved):
Thanks for getting these in here. 📜✨ I did have some questions/suggestions below for your input/consideration.

hillaryj commented on README.md:
````diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+### Using Docker ###
 
-Pull `cisagov/example` from the Docker repository:
+```console
````
I do recommend adding context around command-line instructions so that folks without much experience understand the purpose. Something like:

- To run the included `cisagov/example` image via Docker:

jsf9k replied:
That is a good point, although we devs tend to have a hard time knowing or remembering what normies know and don't know.  It's definitely something to keep in mind when reviewing other devs' pull requests.

This triggered a thought:  Should we have a document in our dev guide that talks about reviewing PRs and has a list of thangs like this to keep in mind?  I think such a document could be useful, but could also turn into a list that is too long and random to be of any use.  Does 18F or TTS have a document we can borrow/steal?

hillaryj replied:
That's a good thought - we could also add an entry to the style guide about "add context around `console` and command-line instructions".

I bet someone's made a document like that of "how to keep your documentation accessible and newbie friendly" but I don't think I've seen one from 18F/TTS.

hillaryj commented on README.md:
````diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+```
 
-    docker pull cisagov/example
+## Using a Docker composition ###
````
```suggestion
### Running with Docker Compose ###
```

hillaryj replied:
This displays at the level above the `Docker` section above, so this both fixes the header level and mirrors later header usage.

hillaryj commented on README.md:
```diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+## Using a Docker composition ###
 
-Or build `cisagov/example` from source:
+1. Create a `docker-compose.yml` file similar to the one below.
```
```suggestion
1. Create a `docker-compose.yml` file similar to the one below to use [Docker Compose](https://docs.docker.com/compose/).
```

hillaryj commented on README.md:
````diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+    ```
 
-This container exposes the following ports:
+## Using secrets ##
````
```suggestion
## Using secrets with your container ##
```

hillaryj replied:
Also, should this still be in the `Running` h2 section (and move to h3 itself) or is it intended to be a top-level h2 section?

hillaryj commented on README.md:
````diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+            target: quote.txt
+    ```
+
+## Updating ##
````
```suggestion
## Updating your container ##
```

hillaryj commented on README.md:
```diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+
+## Updating ##
+
+### Docker-compose ###
```
Switch the orders here - the order of the `Docker` and `Docker Compose` sections should align with their order above.

hillaryj commented on README.md:
```diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+## Running ##
 
-### Install ###
+### Using Docker ###
```
```suggestion
### Running with Docker ###
```

hillaryj commented on README.md:
````diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+    docker pull cisagov/example:0.0.1
+    ```
+
+1. Follow the previous instructions for [running](#running) the container above.
````
```suggestion
1. Recreate and run the container by following the [previous instructions](#running-with-docker).
```

hillaryj commented on README.md:
```diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+
+## Cross-platform builds ##
+
+To create images that are compatible with other platforms you can use the
```
```suggestion
To create images that are compatible with other platforms, you can use the
```

hillaryj commented on README.md:
````diff
@@ -17,49 +17,205 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+      --tag cisagov/example:0.0.1 .
+    ```
 
 ## New Repositories from a Skeleton ##
````
```suggestion
## New repositories from a skeleton ##
```

mcdonnnj replied:
I would like this fixed [upstream](https://github.com/cisagov/skeleton-generic/blob/develop/README.md#new-repositories-from-a-skeleton) instead.

hillaryj reviewed (commented):
Looking good! Here are some nits.

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

felddy committed:
Update README.md

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

felddy committed:
Unquote port integers as this is now a parse error in Docker.

felddy committed:
Change compose specification version to match pyenv testing composition.

The version attribute has been depricated and will be removed once our 
systems move to a more recent release.

felddy committed:
Add context around command-line instructions for non-technical folks.

felddy committed:
Update README.md

Co-authored-by: Hillary <hillary.jeffrey@trio.dhs.gov>

jsf9k commented on README.md:
```diff
@@ -17,51 +17,211 @@ information](LICENSE), as well as [pre-commit hooks](https://pre-commit.com)
+
+The images of this container are tagged with [semantic
+versions](https://semver.org) of the underlying example project that they
+containerize.  It is recommended that most users use a version tag.  e.g.,
+`:0.0.1`
```
```suggestion
containerize.  It is recommended that most users use a version tag (e.g.
`:0.0.1`).
```

jsf9k reviewed (approved):
Still approved, although I made one small suggestion that you can take or leave as you see fit.

dav3r reviewed (approved):
Still approved here also. 👍

felddy committed:
Update README.md

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

mcdonnnj committed:
Merge branch 'develop' into improvement/readme_build_instructions
